## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [205]:
import pandas as pd
from pathlib import Path

client_df_path = Path('Resources/client_dataset.csv')

client_df = pd.read_csv(client_df_path)

In [206]:
# Explore data - Get first 5 rows
client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [207]:
client_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [208]:
# Explore data - View the column names in the data
client_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [209]:
# Explore data - Use the describe function to gather some basic statistics
client_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [210]:
# Explore data - Get number of records in the dataframe
client_df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [211]:
# Explore data - Get number of rows and columns for data
client_df.shape


(54639, 18)

In [212]:
# Explore data - Find null values
client_df.isnull().sum()

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [213]:
# Explore data
# Reduce the number of columns to just show categories
client_df_cats = client_df["category"]
client_df_cats.head()

0          decor
1    consumables
2       software
3    consumables
4    consumables
Name: category, dtype: object

In [214]:
# General exploration
# Reduce the number of columns to just show subcategories

client_df_subcats = client_df["subcategory"]
client_df_subcats.head()

0              wall art
1                  pens
2    project management
3                  pens
4                  misc
Name: subcategory, dtype: object

In [215]:
# General exploration of the "category" column
client_df_cats.describe()

count           54639
unique              5
top       consumables
freq            23538
Name: category, dtype: object

In [216]:
# General exploration of the "subcategory" column
client_df_subcats.describe()

count                 54639
unique                   18
top       bathroom supplies
freq                   6424
Name: subcategory, dtype: object

In [217]:
# General exploration - counts for the "category" column
client_df_cats.value_counts()


category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [218]:
# General exploration - counts for the "subcategory" column
client_df_subcats.value_counts()

subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
seating               2664
pens                  2603
logistics             1122
telecom                978
audio visual           732
cabling                655
desks                  510
computers              467
Name: count, dtype: int64

In [219]:
# What three item categories had the most entries?
client_top_cat = client_df_cats.value_counts()
client_top_cat_df = pd.DataFrame(client_top_cat)
client_top_cat_df.head(3)

count
category          
consumables  23538
furniture    11915
software      8400

In [220]:
# For the category with the most entries, which subcategory had the most entries?
client_top_subcats = client_df[['category', 'subcategory']].value_counts()
client_top_subcats.head(1)


category     subcategory      
consumables  bathroom supplies    6424
Name: count, dtype: int64

In [221]:
# Which five clients had the most entries in the data?
client_df_clients = client_df[['first', 'last', 'client_id']].value_counts().head(5)
client_df_clients

first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [222]:
# Store the client ids of those top 5 clients in a list.
client_ids_list = client_df_clients.reset_index()['client_id'].tolist()
client_ids_list


[33615, 66037, 46820, 24741, 38378]

In [223]:
# What was the most common subcategory for each of the top 5 clients?
client_top_subcats = client_df[['client_id', 'subcategory']].value_counts()
client_top_subcats_df = client_top_subcats.reset_index()
client_top_subcats_df.head()

# I feel that I am missing something related to using the list of client IDs\
# to filter the data (the list from the preceding step, but maybe I'm not)


client_id        subcategory  count
0      91091  bathroom supplies     32
1      37715  bathroom supplies     30
2      66037  bathroom supplies     29
3      43607               misc     29
4      84156     paper products     29

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [224]:
# Drop some unnecessary data
orders_df = client_df.copy()
orders_df = orders_df.drop(columns=['first', 'job', 'last', 'email', 'phone'])
orders_df.head()

client_id  order_id  order_date  order_week  order_year         item_id  \
0      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   

      category         subcategory  unit_price  unit_cost  unit_weight  qty  \
0        decor            wall art     1096.80     762.71         7.50  105   
1  consumables                pens       24.95      15.09         1.49   21   
2     software  project management       13.52       7.86         1.68   39   
3  consumables                pens       36.42      24.85         1.23   29   
4  consumables                misc      195.10     108.17        46.43   20   

   line_number  
0            1  
1            0  
2            6  
3            3  
4            1

In [225]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
orders_df['subtotal'] = orders_df['unit_price'] * orders_df['qty']
orders_df.head()


client_id  order_id  order_date  order_week  order_year         item_id  \
0      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   

      category         subcategory  unit_price  unit_cost  unit_weight  qty  \
0        decor            wall art     1096.80     762.71         7.50  105   
1  consumables                pens       24.95      15.09         1.49   21   
2     software  project management       13.52       7.86         1.68   39   
3  consumables                pens       36.42      24.85         1.23   29   
4  consumables                misc      195.10     108.17        46.43   20   

   line_number   subtotal  
0            1  115164.00  
1            0     523.95  
2            6     527.28  
3            3    1056.18  
4            1    3902.00

In [226]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
calc_shipping = lambda x: 7 * x if x > 50 else 10 * x
orders_df['shipping'] = orders_df['unit_weight'].apply(calc_shipping)
orders_df.head()

client_id  order_id  order_date  order_week  order_year         item_id  \
0      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   

      category         subcategory  unit_price  unit_cost  unit_weight  qty  \
0        decor            wall art     1096.80     762.71         7.50  105   
1  consumables                pens       24.95      15.09         1.49   21   
2     software  project management       13.52       7.86         1.68   39   
3  consumables                pens       36.42      24.85         1.23   29   
4  consumables                misc      195.10     108.17        46.43   20   

   line_number   subtotal  shipping  
0            1  115164.00      75.0  
1            0     523.95      14.9  
2            6     527.28      16.8  
3            3    1056.18      12.3  
4            1    3902.00     464.3

In [227]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
orders_df['total'] = orders_df['subtotal'] + orders_df['shipping']
orders_df.head()

client_id  order_id  order_date  order_week  order_year         item_id  \
0      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   

      category         subcategory  unit_price  unit_cost  unit_weight  qty  \
0        decor            wall art     1096.80     762.71         7.50  105   
1  consumables                pens       24.95      15.09         1.49   21   
2     software  project management       13.52       7.86         1.68   39   
3  consumables                pens       36.42      24.85         1.23   29   
4  consumables                misc      195.10     108.17        46.43   20   

   line_number   subtotal  shipping      total  
0            1  115164.00      75.0  115239.00  
1            0     523.95      14.9     538.85  
2            6     527.28      16.8     544.08  
3            3    1056.18      12.3    1068.48  
4            1    3902.00     464.3    4366.30

In [228]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
orders_df['cost'] = (orders_df['unit_cost'] * orders_df['qty']) + orders_df['shipping']
orders_df.head()


client_id  order_id  order_date  order_week  order_year         item_id  \
0      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   

      category         subcategory  unit_price  unit_cost  unit_weight  qty  \
0        decor            wall art     1096.80     762.71         7.50  105   
1  consumables                pens       24.95      15.09         1.49   21   
2     software  project management       13.52       7.86         1.68   39   
3  consumables                pens       36.42      24.85         1.23   29   
4  consumables                misc      195.10     108.17        46.43   20   

   line_number   subtotal  shipping      total      cost  
0            1  115164.00      75.0  115239.00  80159.55  
1            0     523.95      14.9     538.85    331.79  
2            6     527.28      16.8     544.08    323.34  
3            3    1056.18      12.3    1068.48    732.95  
4            1    3902.00     464.3    4366.30   2627.70

In [229]:
# Create a column for the profit of each line using line cost and line price
orders_df['profit'] = orders_df['total'] - orders_df['cost']
orders_df.head()


client_id  order_id  order_date  order_week  order_year         item_id  \
0      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   

      category         subcategory  unit_price  unit_cost  unit_weight  qty  \
0        decor            wall art     1096.80     762.71         7.50  105   
1  consumables                pens       24.95      15.09         1.49   21   
2     software  project management       13.52       7.86         1.68   39   
3  consumables                pens       36.42      24.85         1.23   29   
4  consumables                misc      195.10     108.17        46.43   20   

   line_number   subtotal  shipping      total      cost    profit  
0            1  115164.00      75.0  115239.00  80159.55  35079.45  
1            0     523.95      14.9     538.85    331.79    207.06  
2            6     527.28      16.8     544.08    323.34    220.74  
3            3    1056.18      12.3    1068.48    732.95    335.53  
4            1    3902.00     464.3    4366.30   2627.70   1738.60

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [233]:
# Check your work using the totals above

# Use a function to calculate the total price for a full order
def calculate_and_print_order_total(orders_df, order_id):
    """
    Calculate and print the total, including shipping and tax, for a full order. 
    Parameters:
    - orders_df: DataFrame containing order data.
    - order_id: The ID of the order to calculate the total price for.
    """
    # Filter the orders_df DataFrame for the specific order_id
    filtered_order_df = orders_df[orders_df['order_id'] == order_id]

    # Sum the 'total' column for the filtered DataFrame
    total_order_price = filtered_order_df['total'].sum()

    # Print the result
    print(f"The total price for Order ID {order_id} is: ${total_order_price:,.2f}")

# Order ID 2742071 had a total price of $152,811.89
calculate_and_print_order_total(orders_df, 2742071)

# Order ID 2173913 had a total price of $162,388.71
calculate_and_print_order_total(orders_df, 2173913)

# Order ID 6128929 had a total price of $923,441.25
calculate_and_print_order_total(orders_df, 6128929)

# Summary: These do NOT match. I will need to go back and check my work. I suspect that \
# I may have made a mistake in the calculations.


The total price for Order ID 2742071 is: $128,004.13
The total price for Order ID 2173913 is: $109,691.86
The total price for Order ID 6128929 is: $689,014.87


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [234]:
# I will go back and check my work to see if I can find the error. I had to stop the 
# the assignment here because I was running out of time. My apologies for not being able to
# resolve the issue. I will continue to work on this assignment and will resubmit 
# it as soon as I can.

# I also ran into an issue where my `git stash` was corrupted and I had to reset 
# my local git history. I lost a lot of work that I had done on this 
# assignment as a result. An unfortunate situation, but I will continue to 
# work on this assignment and will resubmit it as soon as I can.

# I know being able to use groupby or pivot_table would be immensely helpful in this
# situation, but without the ability to use those functions, I am struggling to solve
# the problem. 



# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Use the client_ids_list from the previous step to filter the data



client_id  order_id  order_date  order_week  order_year         item_id  \
5        33615   1327730  2023-05-02          18        2023  BMB23158-88-9A   
12       33615   6059361  2023-05-16          20        2023  BDC65078-60-7M   
137      46820   7109502  2023-05-12          19        2023  CXU60946-73-5B   
214      46820   8967498  2023-02-06           6        2023  EMX21429-07-6D   
216      66037   1203693  2023-04-09          14        2023  XUX09951-98-0M   

        category        subcategory  unit_price  unit_cost  unit_weight  qty  \
5          decor             plants      172.84     154.44         2.79   78   
12   consumables     paper products      229.74     198.38        27.51  568   
137  consumables  bathroom supplies      136.13     111.47         5.65  118   
214    furniture             tables        7.98       6.26         7.05  133   
216  consumables  bathroom supplies      301.87     166.09         0.26  157   

     line_number   subtotal  shipping      total       cost    profit  
5              0   13481.52      27.9   13509.42   12074.22   1435.20  
12             2  130492.32     275.1  130767.42  112954.94  17812.48  
137            9   16063.34      56.5   16119.84   13209.96   2909.88  
214            3    1061.34      70.5    1131.84     903.08    228.76  
216            2   47393.59       2.6   47396.19   26078.73  21317.46

In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [ ]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
